src/chain/unzip.rs

```diff
@@ -29,174 +29,82 @@ pub trait Unzippable: Sized {
-        result
-    }
-
-    fn make_targets(commands: &mut Commands) -> SmallVec<[Entity; 8]> {
-        SmallVec::from_iter([
-            commands.spawn(UnusedTarget).id(),
-            commands.spawn(UnusedTarget).id(),
-        ])
-    }
-
-    fn distribute_values(
-        OperationRequest { source, world, roster }: OperationRequest,
-    ) -> OperationResult {
-        let Input { session, data: inputs } = world
-            .get_entity_mut(source).or_broken()?
-            .take_input::<Self>()?;
-
-        let targets = world.get::<ForkTargetStorage>(source).or_broken()?;
-        let target_0 = *targets.0.get(0).or_broken()?;
-        let target_1 = *targets.0.get(1).or_broken()?;
-
-        if let Some(mut t_mut) = world.get_entity_mut(target_0) {
-            t_mut.give_input(session, inputs.0, roster)?;
-        }
-
-        if let Some(mut t_mut) = world.get_entity_mut(target_1) {
-            t_mut.give_input(session, inputs.1, roster)?;
+macro_rules! impl_unzippable_for_tuple {
+    ($($T:ident),*) => {
+        #[allow(non_snake_case)]
+        impl<$($T: 'static + Send + Sync),*> Unzippable for ($($T,)*)
+        {
+            type Unzipped = ($(Output<$T>,)*);
+            fn unzip_output(output: Output<Self>, builder: &mut Builder) -> Self::Unzipped {
+                assert_eq!(output.scope(), builder.scope());
+                let mut targets = SmallVec::new();
+                let result =
+                    (
+                        $(
+                            {
+                                // Variable is only used to make sure this cycle is repeated once
+                                // for each instance of the $T type, but the type itself is not
+                                // used.
+                                #[allow(unused)]
+                                let $T = std::marker::PhantomData::<$T>;
+                                let target = builder.commands.spawn(UnusedTarget).id();
+                                targets.push(target);
+                                Output::new(builder.scope, target)
+                            },
+                        )*
+                    );
+
+                builder.commands.add(AddOperation::new(
+                    Some(output.scope()),
+                    output.id(),
+                    ForkUnzip::<Self>::new(ForkTargetStorage(targets)),
+                ));
+                result
+            }
+
+            fn distribute_values(
+                OperationRequest { source, world, roster }: OperationRequest,
+            ) -> OperationResult {
+                let Input { session, data: inputs } = world
+                    .get_entity_mut(source).or_broken()?
+                    .take_input::<Self>()?;
+
+                // Targets is cloned to avoid borrow checker issues when
+                // doing a mutable borrow of the world later
+                let targets = world.get::<ForkTargetStorage>(source).or_broken()?.clone();
```

Comment:
From my understanding previously the clone wasn't necessary because we would dereference the borrow to `Entity` which, because `Entity` implements `Copy`, would just copy it below the scenes.
Performance wise I believe cloning the `SmallVec` should be _almost_ the same as copying all the elements one by one

Reply:
Right, the problem here is we can't guarantee that the preallocated `SmallVec` size is big enough since it's currently [8](https://github.com/open-rmf/bevy_impulse/blob/main/src/operation.rs#L166), and these tuples go up to 12. At that point doing a clone will cause an avoidable heap allocation.

If there's any way to use a pattern here to create `N` variables with entities or to create a tuple of `N` entities, copied out of the `SmallVec`, that would be preferable. Heap allocations once in a while aren't a huge deal if avoiding them is too onerous, but the more we can avoid them inside of hot loops, the more compelling this library is.

Reply:
Fair, I "found" a trick that by adding a dummy macro element (that shouldn't really cost anything since it's only a compile time cost) we can unpack all the vector elements into a tuple, which also fairly simplifies the code there https://github.com/open-rmf/bevy_impulse/pull/2/commits/dc95e402ded1e812b0132ece141e038a7b0a3103.
This comes at the cost of a dependency on `itertools` since it has a very convenient method to turn an iterator into a tuple.

I had to change the maximum tuple size down to 12 since that's the maximum that itertools `next_tuple()` implements.

Reply:
Awesome!

> I had to change the maximum tuple size down to 12 since that's the maximum that itertools next_tuple() implements.

This should be totally fine. Since it supports nested tuples of any depth, there's effectively no limit, but if people are exceeding tuples of 12 elements in one output then they're going pretty wild anyway.